assets/product-info.js

```diff
@@ -68,37 +69,30 @@ if (!customElements.get('product-info')) {
         });
       }
 
-      handleOptionValueChange({ data: { event, target, variant } }) {
+      handleOptionValueChange({ data: { event, target, selectedOptionValues } }) {
         if (!this.contains(event.target)) return;
 
-        const targetUrl = target.dataset.productUrl || this.dataset.url;
+        const baseUrl = target.dataset.productUrl || this.inflightRequestProductUrl || this.dataset.url;
+        this.inflightRequestProductUrl = baseUrl;
```

Comment:
I think I'm missing something. Why do we need the `inflightRequestProductUrl` ?

- Its value is defined on variant selection change. 
- We then run the fetch request based on the `baseUrl`
- When the fetch runs, it will set the `inflightRequestProductUrl` back to `null` if it has completed. 
- If you happen to change the variant selection while a fetch is running, it will go through `handleOptionValueChange` and update the `baseUrl` in case it has a product url associated with it, if not it will use the `inflightRequestProductUrl`. But why is that needed ? Maybe because I'm testing on my store without combined listing items. 

As a user I click on a variant which happens to have a `data-product-url` associated to it, so it: 
- sets the `baseUrl` and `inflightRequestProductUrl` to the same value
- runs the fetch...
- but I change variant selection while doing so, and re sets the base url to either the `data-product-url` of the option I clicked on or the previous baseUrl value via `this.inflightRequestProductUrl`. But what if that option I clicked on doesn't have a `data-product-url` (null) ? Wouldn't we be using the wrong `baseUrl` then ?

Reply:
> But what if that option I clicked on doesn't have a data-product-url (null) ? Wouldn't we be using the wrong baseUrl then

Yeah exactly, this is the case that `inflightRequestProductUrl` is meant to account for. Only parent options have product urls:

<img width="652" alt="04-31-4oqnl-q7v6u" src="https://github.com/Shopify/dawn/assets/3619012/c24bfb9e-d434-43fc-8adb-6e5360b82e86">

Without this change, the bug that presents is that the final request uses stale information, and can return the wrong product or the wrong option value selection. For example:
- Buyer starts with a selection of `black/m/simple`
- 👍  Buyer clicks Orange, sfapi is called with `/pants-orange?option_values=<orange_id>,<m_id>,<simple_id>`
- 👎  _Before_ the first request resolves, buyer clicks `style: fancy`. Because there is no product-url, the code used to assume this meant the option value change was within the currently active product. Also, because option values haven't been updated yet from the first request response, the variant associated with the option value is stale. So here, having `orange/m/simple` selected and clicking `fancy` fires a request for `/pants-black?variant=<black/m/fancy>` but it should request for `/pants-orange?option_values=<orange_id>,<m_id>,<fancy_id>`

With the changes, we do 2 things to make sure we request the right product:
1) Cache `inflightRequestProductUrl` after each request so we can fallback to the correct parent product selection if product-url is null
2) Call the api with option_values instead of variant

So with the same example:
- Buyer starts with a selection of `black/m/simple`
- 👍  Buyer clicks Orange, sfapi is called with `/pants-orange?option_values=<orange_id>,<m_id>,<simple_id>`
- 👍  Set `this.inflightRequestProductUrl = /pants-orange`
- _Before_ the first request resolves, buyer clicks `style: fancy`
  - 👍 This fallback `const baseUrl = target.dataset.productUrl || this.inflightRequestProductUrl || ...` returns the correct parent product url of `/pants-orange`
  - 👍 [We read the selected option values](https://github.com/Shopify/dawn/pull/3495/files#r1621561196) to get `/pants-orange?option_values=<orange_id>,<m_id>,<fancy_id>`

Reply:
> Maybe because I'm testing on my store without combined listing items.

Yeah that's definitely it, this weird case only comes up with combined listings 👍

Reply:
I'm noticing an issue with the complementary block within the product info on `main-product.liquid`. 
When switching variants in that modal, the url it's using for the fetch request is sometimes wrong and breaks the experience: [recording](https://screenshot.click/05-55-9v5s6-diow0.mp4)

Reply:
PR here: https://github.com/Shopify/dawn/pull/3504